File: 2025/239. Sliding Window Maximum/solution.py
class MonotonicQueue:
    def __init__(self):
        self.queue = deque()
    
    def push(self, val):
        while self.queue and self.queue[-1] < val:
            self.queue.pop()
        self.queue.append(val)
    
    def pop(self, val):
        if self.queue and self.queue[0] == val:
            self.queue.popleft()

    def max(self):
        return self.queue[0] if self.queue else None

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        queue = MonotonicQueue()
        res = []
        l = r = 0
        while r < len(nums):
            queue.push(nums[r])
            if r - l + 1 >= k:
                res.append(queue.max())
                queue.pop(nums[l])
                l += 1
            r += 1
        return res

File: 2025/239. Sliding Window Maximum/Solution.java
class Solution {
    public int[] maxSlidingWindow(int[] nums, int k) {
        Deque<Integer> monoDecreasingQueue = new ArrayDeque<>();
        int[] res = new int[nums.length - k + 1];
        int j = 0;
        for (int i = 0; i < nums.length; i++) {
            while (!monoDecreasingQueue.isEmpty() && monoDecreasingQueue.peekLast() < nums[i]) {
                    monoDecreasingQueue.pollLast();
            }
            monoDecreasingQueue.offerLast(nums[i]);
            if (i >= k - 1) {
                res[j++] = monoDecreasingQueue.peekFirst();
                if (!monoDecreasingQueue.isEmpty() && monoDecreasingQueue.peekFirst() == nums[i - k + 1]) {
                    monoDecreasingQueue.pollFirst();
                }
            }
        }
        return res;
    }
}